File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sharing/SharedLocationLinkResolver.kt
package expo.modules.vescapecore.sharing

import java.net.URI
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request

data class ResolvedSharedLocation(
  val latitude: Double,
  val longitude: Double,
  val name: String?,
) {
  fun toBridgeMap(): Map<String, Any?> = mapOf(
    "latitude" to latitude,
    "longitude" to longitude,
    "name" to name,
  )
}

internal data class SharedLocationHttpResponse(
  val url: String,
  val location: String?,
  val body: String,
  val status: Int = 200,
)

/**
 * Resolves opaque map share links without a Maps API. Google uses different redirect chains for
 * Android and iOS shares, so redirects are deliberately stepped before the final browser-like GET.
 *
 * @parity /modules/vescape-core/ios/sharing/SharedLocationLinkResolver.swift
 * @parity /modules/vescape-core/src/index.ts `resolveSharedLocationLink`
 */
class SharedLocationLinkResolver internal constructor(
  private val request: suspend (String, Boolean, String, String?) -> SharedLocationHttpResponse,
) {
  constructor() : this(defaultRequest())

  suspend fun resolve(link: String): ResolvedSharedLocation? {
    var current = link
    repeat(MAX_REDIRECTS) {
      var response = request(current, false, SIMPLE_USER_AGENT, null)
      var attempts = 1
      while (response.status == 404 && attempts < MAX_RETRIES) {
        delay(RETRY_DELAY_MS)
        response = request(current, false, SIMPLE_USER_AGENT, null)
        attempts += 1
      }
      val destination = response.location?.let { resolveRedirect(current, it) }
        ?: return extract(current, response.body)
      current = destination
    }

    val consentDestination = googleConsentDestination(current)
    if (consentDestination != null) current = consentDestination
    val response = request(
      current,
      true,
      BROWSER_USER_AGENT,
      if (consentDestination != null) GOOGLE_CONSENT_COOKIE else null,
    )
    return extract(response.url, response.body)
  }

  companion object {
    private const val MAX_REDIRECTS = 2
    private const val MAX_RETRIES = 5
    private const val RETRY_DELAY_MS = 250L
    private const val GOOGLE_CONSENT_COOKIE = "SOCS=CAESHAgBEhIaAB; CONSENT=YES+"
    private const val SIMPLE_USER_AGENT = "Mozilla/5.0"
    private const val BROWSER_USER_AGENT =
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
        "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"

    private val PIN_PATTERN = Regex("!3d(-?\\d+(?:\\.\\d+)?)!4d(-?\\d+(?:\\.\\d+)?)")
    private val ROUTE_PATTERN = Regex("!\\d+d(-?\\d+(?:\\.\\d+)?)!\\d+d(-?\\d+(?:\\.\\d+)?)")
    private val INITIAL_STATE_PATTERN = Regex(
      "window\\.APP_INITIALIZATION_STATE=\\[\\[\\[-?\\d+(?:\\.\\d+)?,\\s*" +
        "(-?\\d+(?:\\.\\d+)?),\\s*(-?\\d+(?:\\.\\d+)?)",
    )
    private val STATIC_CENTER_PATTERN = Regex(
      "(?:center=|center%3D)(-?\\d+(?:\\.\\d+)?)(?:%2C|,)(-?\\d+(?:\\.\\d+)?)",
      RegexOption.IGNORE_CASE,
    )
    private val QUERY_PATTERN = Regex(
      "[?&](?:q|query|ll|sll|daddr|saddr|destination|center|coordinate)=(?:loc:)?" +
        "(-?\\d+(?:\\.\\d+)?)(?:%2C|,)(-?\\d+(?:\\.\\d+)?)",
      RegexOption.IGNORE_CASE,
    )
    private val OSM_MARKER_PATTERN = Regex(
      "[?&]mlat=(-?\\d+(?:\\.\\d+)?)(?:&|&amp;)mlon=(-?\\d+(?:\\.\\d+)?)",
      RegexOption.IGNORE_CASE,
    )
    private val VIEWPORT_PATTERN = Regex(
      "@(-?\\d+(?:\\.\\d+)?)(?:%2C|,)(-?\\d+(?:\\.\\d+)?)",
      RegexOption.IGNORE_CASE,
    )

    internal fun extract(url: String, body: String): ResolvedSharedLocation? {
      coordinateMatch(PIN_PATTERN.find(url), reversed = false)?.let { return it.withName(url) }
      coordinateMatch(ROUTE_PATTERN.find(url), reversed = false)?.let { return it.withName(url) }
      coordinateMatch(OSM_MARKER_PATTERN.find(url), reversed = false)?.let { return it.withName(url) }
      coordinateMatch(QUERY_PATTERN.find(url), reversed = false)?.let { return it.withName(url) }
      coordinateMatch(VIEWPORT_PATTERN.find(url), reversed = false)?.let { return it.withName(url) }
      coordinateMatch(INITIAL_STATE_PATTERN.find(body), reversed = true)?.let { return it.withName(url) }
      coordinateMatch(STATIC_CENTER_PATTERN.find(body), reversed = false)?.let { return it.withName(url) }
      return null
    }

    private fun coordinateMatch(match: MatchResult?, reversed: Boolean): ResolvedSharedLocation? {
      val first = match?.groupValues?.getOrNull(1)?.toDoubleOrNull() ?: return null
      val second = match.groupValues.getOrNull(2)?.toDoubleOrNull() ?: return null
      val latitude = if (reversed) second else first
      val longitude = if (reversed) first else second
      if (latitude !in -90.0..90.0 || longitude !in -180.0..180.0) return null
      return ResolvedSharedLocation(latitude, longitude, null)
    }

    private fun ResolvedSharedLocation.withName(url: String): ResolvedSharedLocation =
      copy(name = Regex("/maps/place/([^/@?#]+)", RegexOption.IGNORE_CASE).find(url)
        ?.groupValues?.getOrNull(1)
        ?.replace('+', ' ')
        ?.let { runCatching { java.net.URLDecoder.decode(it, Charsets.UTF_8.name()) }.getOrDefault(it) }
        ?.trim()
        ?.takeIf(String::isNotEmpty))

    private fun resolveRedirect(base: String, location: String): String =
      runCatching { URI(base).resolve(location).toString() }.getOrDefault(location)

    private fun googleConsentDestination(url: String): String? = runCatching {
      val uri = URI(url)
      if (!uri.host.equals("consent.google.com", ignoreCase = true)) return null
      uri.rawQuery
        ?.split('&')
        ?.firstOrNull { it.substringBefore('=') == "continue" }
        ?.substringAfter('=', "")
        ?.let { java.net.URLDecoder.decode(it, Charsets.UTF_8.name()) }
        ?.takeIf(String::isNotEmpty)
    }.getOrNull()

    private fun defaultRequest(): suspend (String, Boolean, String, String?) -> SharedLocationHttpResponse {
      val followingClient = OkHttpClient.Builder().followRedirects(true).build()
      val blockingClient = OkHttpClient.Builder().followRedirects(false).build()
      return { url, followRedirects, userAgent, cookie ->
        withContext(Dispatchers.IO) {
          val requestBuilder = Request.Builder().url(url).header("User-Agent", userAgent)
          if (cookie != null) requestBuilder.header("Cookie", cookie)
          val call = (if (followRedirects) followingClient else blockingClient).newCall(
            requestBuilder.build(),
          )
          call.execute().use { response ->
            SharedLocationHttpResponse(
              url = response.request.url.toString(),
              location = response.header("Location"),
              body = response.body?.string().orEmpty(),
              status = response.code,
            )
          }
        }
      }
    }
  }
}

File: modules/vescape-core/ios/sharing/SharedLocationLinkResolver.swift
import Foundation

struct ResolvedSharedLocation {
  let latitude: Double
  let longitude: Double
  let name: String?

  var bridgeValue: [String: Any?] {
    ["latitude": latitude, "longitude": longitude, "name": name]
  }
}

struct SharedLocationHTTPResponse {
  let url: String
  let location: String?
  let body: String
  let status: Int

  init(url: String, location: String?, body: String, status: Int = 200) {
    self.url = url
    self.location = location
    self.body = body
    self.status = status
  }
}

private final class RedirectBlockingDelegate: NSObject, URLSessionTaskDelegate {
  func urlSession(
    _: URLSession,
    task _: URLSessionTask,
    willPerformHTTPRedirection _: HTTPURLResponse,
    newRequest _: URLRequest,
    completionHandler: @escaping (URLRequest?) -> Void
  ) {
    completionHandler(nil)
  }
}

/**
 Resolves opaque map share links without a Maps API. Google uses different redirect chains for
 Android and iOS shares, so redirects are deliberately stepped before the final browser-like GET.

 @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/sharing/SharedLocationLinkResolver.kt
 @parity /modules/vescape-core/src/index.ts `resolveSharedLocationLink`
 */
final class SharedLocationLinkResolver {
  typealias Request = (String, Bool, String, String?) async throws -> SharedLocationHTTPResponse

  private let request: Request

  init(request: @escaping Request = SharedLocationLinkResolver.defaultRequest) {
    self.request = request
  }

  func resolve(_ link: String) async throws -> ResolvedSharedLocation? {
    var current = link
    for _ in 0..<2 {
      var response = try await request(current, false, Self.simpleUserAgent, nil)
      var attempts = 1
      while response.status == 404 && attempts < 5 {
        try await Task.sleep(for: .milliseconds(250))
        response = try await request(current, false, Self.simpleUserAgent, nil)
        attempts += 1
      }
      guard let location = response.location,
            let destination = URL(string: location, relativeTo: URL(string: current))?.absoluteURL.absoluteString
      else {
        return Self.extract(url: current, body: response.body)
      }
      current = destination
    }

    let consentDestination = Self.googleConsentDestination(current)
    if let consentDestination { current = consentDestination }
    let response = try await request(
      current,
      true,
      Self.browserUserAgent,
      consentDestination == nil ? nil : Self.googleConsentCookie
    )
    return Self.extract(url: response.url, body: response.body)
  }

  static func extract(url: String, body: String) -> ResolvedSharedLocation? {
    if let coordinate = coordinate(in: url, pattern: pinPattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: url, pattern: routePattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: url, pattern: osmMarkerPattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: url, pattern: queryPattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: url, pattern: viewportPattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: body, pattern: initialStatePattern, reversed: true) {
      return coordinate.withName(from: url)
    }
    if let coordinate = coordinate(in: body, pattern: staticCenterPattern, reversed: false) {
      return coordinate.withName(from: url)
    }
    return nil
  }

  private static let simpleUserAgent = "Mozilla/5.0"
  private static let googleConsentCookie = "SOCS=CAESHAgBEhIaAB; CONSENT=YES+"
  private static let browserUserAgent =
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
    "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
  private static let pinPattern = #"!3d(-?\d+(?:\.\d+)?)!4d(-?\d+(?:\.\d+)?)"#
  private static let routePattern = #"!\d+d(-?\d+(?:\.\d+)?)!\d+d(-?\d+(?:\.\d+)?)"#
  private static let initialStatePattern =
    #"window\.APP_INITIALIZATION_STATE=\[\[\[-?\d+(?:\.\d+)?,\s*(-?\d+(?:\.\d+)?),\s*(-?\d+(?:\.\d+)?)"#
  private static let staticCenterPattern =
    #"(?:center=|center%3D)(-?\d+(?:\.\d+)?)(?:%2C|,)(-?\d+(?:\.\d+)?)"#
  private static let queryPattern =
    #"[?&](?:q|query|ll|sll|daddr|saddr|destination|center|coordinate)=(?:loc:)?(-?\d+(?:\.\d+)?)(?:%2C|,)(-?\d+(?:\.\d+)?)"#
  private static let osmMarkerPattern =
    #"[?&]mlat=(-?\d+(?:\.\d+)?)(?:&|&amp;)mlon=(-?\d+(?:\.\d+)?)"#
  private static let viewportPattern = #"@(-?\d+(?:\.\d+)?)(?:%2C|,)(-?\d+(?:\.\d+)?)"#

  private static func coordinate(
    in value: String,
    pattern: String,
    reversed: Bool
  ) -> ResolvedSharedLocation? {
    guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]),
          let match = regex.firstMatch(in: value, range: NSRange(value.startIndex..., in: value)),
          let firstRange = Range(match.range(at: 1), in: value),
          let secondRange = Range(match.range(at: 2), in: value),
          let first = Double(value[firstRange]),
          let second = Double(value[secondRange])
    else { return nil }
    let latitude = reversed ? second : first
    let longitude = reversed ? first : second
    guard (-90...90).contains(latitude), (-180...180).contains(longitude) else { return nil }
    return ResolvedSharedLocation(latitude: latitude, longitude: longitude, name: nil)
  }

  private static func defaultRequest(
    _ url: String,
    _ followRedirects: Bool,
    _ userAgent: String,
    _ cookie: String?
  ) async throws -> SharedLocationHTTPResponse {
    guard let requestURL = URL(string: url) else { throw URLError(.badURL) }
    var urlRequest = URLRequest(url: requestURL)
    urlRequest.setValue(userAgent, forHTTPHeaderField: "User-Agent")
    if let cookie { urlRequest.setValue(cookie, forHTTPHeaderField: "Cookie") }
    let session = followRedirects
      ? URLSession.shared
      : URLSession(configuration: .ephemeral, delegate: RedirectBlockingDelegate(), delegateQueue: nil)
    let (data, response) = try await session.data(for: urlRequest)
    guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
    return SharedLocationHTTPResponse(
      url: http.url?.absoluteString ?? url,
      location: http.value(forHTTPHeaderField: "Location"),
      body: String(decoding: data, as: UTF8.self),
      status: http.statusCode
    )
  }

  private static func googleConsentDestination(_ value: String) -> String? {
    guard let components = URLComponents(string: value),
          components.host?.lowercased() == "consent.google.com"
    else { return nil }
    return components.queryItems?.first(where: { $0.name == "continue" })?.value
  }
}

private extension ResolvedSharedLocation {
  func withName(from url: String) -> ResolvedSharedLocation {
    guard let regex = try? NSRegularExpression(pattern: #"/maps/place/([^/@?#]+)"#, options: [.caseInsensitive]),
          let match = regex.firstMatch(in: url, range: NSRange(url.startIndex..., in: url)),
          let range = Range(match.range(at: 1), in: url)
    else { return self }
    let encoded = String(url[range]).replacingOccurrences(of: "+", with: " ")
    let decoded = encoded.removingPercentEncoding ?? encoded
    return ResolvedSharedLocation(latitude: latitude, longitude: longitude, name: decoded)
  }
}
